src/conn/conn_compact.c

```diff
@@ -0,0 +1,253 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __compact_server_run_chk --
+ *     Check to decide if the compact server should continue running.
+ */
+static bool
+__compact_server_run_chk(WT_SESSION_IMPL *session)
+{
+    return (FLD_ISSET(S2C(session)->server_flags, WT_CONN_SERVER_COMPACT));
+}
+
+/*
+ * __compact_server --
+ *     The compact server thread.
+ */
+static WT_THREAD_RET
+__compact_server(void *arg)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    WT_SESSION *wt_session;
+    WT_SESSION_IMPL *session;
+    int exact;
+    const char *config, *key, *prefix;
+    bool full_iteration, running, signalled;
+
+    session = arg;
+    conn = S2C(session);
+    wt_session = (WT_SESSION *)session;
+    cursor = NULL;
+    config = NULL;
+    prefix = "file:";
+    key = prefix;
+    exact = 0;
+    full_iteration = running = signalled = false;
+
+    WT_STAT_CONN_SET(session, session_background_compact_running, 0);
+
+    for (;;) {
+
+        __wt_spin_lock(session, &conn->background_compact.lock);
+        running = conn->background_compact.running;
+        __wt_spin_unlock(session, &conn->background_compact.lock);
+
+        /* When the entire metadata file has been parsed, take a break or wait until signalled. */
+        if (full_iteration || !running) {
+
+            full_iteration = false;
+            /*
+             * FIXME-WT-11409: Depending on the previous state, we may not want to clear out the
+             * last key used. This could be useful if the server was paused to be resumed later.
+             */
+            key = prefix;
+            /* Check every 10 seconds in case the signal was missed. */
+            __wt_cond_wait(
+              session, conn->background_compact.cond, 10 * WT_MILLION, __compact_server_run_chk);
```

Comment:
If I don't put a timeout, I end up with tests that time out, see an example [here](https://spruce.mongodb.com/task/wiredtiger_ubuntu2004_examples_c_production_disable_static_test_patch_7433082882e93ea378a71abc0127231070a48971_64c770210ae6068a921a872e_23_07_31_08_26_10/logs?execution=0).

You can see that `join` is called by the main thread:
```
[2023/07/31 10:53:36.966] Thread 1 (Thread 0x7f56d173c800 (LWP 401438)):
[2023/07/31 10:53:36.966] #0  __pthread_clockjoin_ex (threadid=140010782402304, thread_return=thread_return@entry=0x0, clockid=clockid@entry=0, abstime=abstime@entry=0x0, block=block@entry=true) at pthread_join_common.c:145
[2023/07/31 10:53:36.966] #1  0x00007f56d1c849f3 in __pthread_join (threadid=<optimized out>, thread_return=thread_return@entry=0x0) at pthread_join.c:24
[2023/07/31 10:53:36.966] #2  0x00007f56d1e822a9 in __wt_thread_join (session=session@entry=0x1d12000, tid=tid@entry=0x16f49e0) at .../wiredtiger/src/os_posix/os_thread.c:92
[2023/07/31 10:53:36.966] #3  0x00007f56d1da4458 in __wt_compact_server_destroy (session=session@entry=0x1d12000) at .../wiredtiger/src/conn/conn_compact.c:190
[2023/07/31 10:53:36.966] #4  0x00007f56d1d97868 in __conn_close (wt_conn=0x16f4000, config=<optimized out>) at .../wiredtiger/src/conn/conn_api.c:1188
[2023/07/31 10:53:36.966] #5  0x0000000000405c22 in main (argc=<optimized out>, argv=<optimized out>) at .../wiredtiger/examples/c/ex_all.c:1291
```

And the compaction server is waiting on the condition variable while thread 1 should have signalled it:

```
[2023/07/31 10:53:36.910] Thread 10 (Thread 0x7f56ccf32700 (LWP 402101)):
[2023/07/31 10:53:36.910] #0  futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x1717ad8) at ../sysdeps/nptl/futex-internal.h:183
[2023/07/31 10:53:36.911] #1  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x1717a88, cond=0x1717ab0) at pthread_cond_wait.c:508
[2023/07/31 10:53:36.911] #2  __pthread_cond_wait (cond=cond@entry=0x1717ab0, mutex=mutex@entry=0x1717a88) at pthread_cond_wait.c:647
[2023/07/31 10:53:36.911] #3  0x00007f56d1e81b76 in __wt_cond_wait_signal (session=session@entry=0x1d15440, cond=0x1717a80, usecs=usecs@entry=0, run_func=run_func@entry=0x0, signalled=signalled@entry=0x7f56ccf31def) at .../wiredtiger/src/os_posix/os_mtx_cond.c:117
[2023/07/31 10:53:36.911] #4  0x00007f56d1da404e in __wt_cond_wait (run_func=0x0, usecs=0, cond=<optimized out>, session=0x1d15440) at .../wiredtiger/src/include/misc_inline.h:19
[2023/07/31 10:53:36.911] #5  __compact_server (arg=0x1d15440) at .../wiredtiger/src/conn/conn_compact.c:65
[2023/07/31 10:53:36.911] #6  0x00007f56d1c83609 in start_thread (arg=<optimized out>) at pthread_create.c:477
[2023/07/31 10:53:36.911] #7  0x00007f56d1b7b133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

I agree that the timeout can be seen as a bit suspicious. In conn_capacity.c`, you can find this interesting [comment](https://github.com/wiredtiger/wiredtiger/blob/develop/src/conn/conn_capacity.c#L95-L98):

```
        /*
         * Wait until signalled but check once per second in case the signal was missed.
         */
        __wt_cond_wait(session, conn->capacity_cond, WT_MILLION, __capacity_server_run_chk);
```

It seems that the timeout is used here for a similar reason.

Reply:
Thanks for the details. Let's go with the sleep option.